File: backend/app/models/crm_custom_objects.py
"""CRM Custom Objects models — user-defined entities, records, and relationships."""
from __future__ import annotations

import uuid

from sqlalchemy import (
    Boolean,
    ForeignKey,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import JSON, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class CustomObjectDefinition(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Schema definition for a user-created custom CRM object."""

    __tablename__ = "crm_custom_object_definitions"

    name: Mapped[str] = mapped_column(String(100), nullable=False)
    label: Mapped[str] = mapped_column(String(200), nullable=False)
    plural_label: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    icon: Mapped[str | None] = mapped_column(String(50), nullable=True)
    fields: Mapped[dict | None] = mapped_column(
        JSON, nullable=True
    )  # [{name, label, type, required, options, default_value}, ...]
    relationships: Mapped[dict | None] = mapped_column(
        JSON, nullable=True
    )  # [{related_entity, relationship_type, foreign_key_field}, ...]
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )

    records = relationship("CustomObjectRecord", back_populates="definition", cascade="all, delete-orphan")
    creator = relationship("User", foreign_keys=[created_by])


class CustomObjectRecord(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Single record (instance) of a custom object."""

    __tablename__ = "crm_custom_object_records"

    definition_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("crm_custom_object_definitions.id"), nullable=False
    )
    data: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    owner_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )

    definition = relationship("CustomObjectDefinition", back_populates="records")
    owner = relationship("User", foreign_keys=[owner_id])
    record_relationships = relationship("CustomObjectRelationship", back_populates="record", cascade="all, delete-orphan")


class CustomObjectRelationship(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    """Link between a custom object record and another entity."""

    __tablename__ = "crm_custom_object_relationships"

    record_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("crm_custom_object_records.id"), nullable=False
    )
    related_entity_type: Mapped[str] = mapped_column(String(50), nullable=False)
    related_entity_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    relationship_type: Mapped[str] = mapped_column(String(50), nullable=False)

    record = relationship("CustomObjectRecord", back_populates="record_relationships")

File: frontend/src/api/crm_custom_objects.ts
/**
 * CRM Custom Objects API client — user-defined entity types and their records.
 *
 * Exports TanStack Query hooks and Axios helper functions for the CRM custom
 * objects feature. All requests go through `client.ts` (Axios instance with
 * auth interceptors). Backend prefix: `/api/v1/crm`.
 *
 * Key exports:
 *   - useCustomObjects() — list all custom object definitions
 *   - useCreateCustomObject() / useUpdateCustomObject() / useDeleteCustomObject() — definition mutations
 *   - useCustomObjectRecords() — list records for a given definition
 *   - useCreateCustomObjectRecord() / useUpdateCustomObjectRecord() — record mutations
 *   - useCustomObjectRelationships() / useCreateRelationship() — link records to CRM entities
 *
 * Note: field schema (CustomObjectField[]) is stored as JSON on each definition.
 * Records hold arbitrary key/value data conforming to that schema.
 */
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import apiClient from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface CustomObjectField {
  name: string
  label: string
  type: string
  required?: boolean
  options?: string[]
  default_value?: unknown
}

export interface CustomObjectDefinition {
  id: string
  name: string
  label: string
  plural_label: string
  description: string | null
  icon: string | null
  fields: CustomObjectField[] | null
  relationships: Record<string, unknown>[] | null
  is_active: boolean
  created_by: string
  created_at: string
  updated_at: string
}

export interface CustomObjectRecord {
  id: string
  definition_id: string
  data: Record<string, unknown> | null
  owner_id: string
  created_at: string
  updated_at: string
}

export interface CustomObjectRelationship {
  id: string
  record_id: string
  related_entity_type: string
  related_entity_id: string
  relationship_type: string
}

// ─── API Functions ────────────────────────────────────────────────────────────

const getCustomObjects = (params?: Record<string, unknown>) =>
  apiClient.get('/crm/custom-objects', { params }).then((r) => r.data)

const createCustomObject = (data: Partial<CustomObjectDefinition>) =>
  apiClient.post('/crm/custom-objects', data).then((r) => r.data)

const getCustomObject = (id: string) =>
  apiClient.get(`/crm/custom-objects/${id}`).then((r) => r.data)

const updateCustomObject = ({ id, ...data }: Partial<CustomObjectDefinition> & { id: string }) =>
  apiClient.put(`/crm/custom-objects/${id}`, data).then((r) => r.data)

const deleteCustomObject = (id: string) =>
  apiClient.delete(`/crm/custom-objects/${id}`).then((r) => r.data)

const getCustomObjectRecords = (objectId: string, params?: Record<string, unknown>) =>
  apiClient.get(`/crm/custom-objects/${objectId}/records`, { params }).then((r) => r.data)

const createRecord = ({ objectId, ...data }: Partial<CustomObjectRecord> & { objectId: string }) =>
  apiClient.post(`/crm/custom-objects/${objectId}/records`, data).then((r) => r.data)

const getCustomObjectRecord = (objectId: string, recordId: string) =>
  apiClient.get(`/crm/custom-objects/${objectId}/records/${recordId}`).then((r) => r.data)

const updateRecord = ({
  objectId,
  recordId,
  ...data
}: Partial<CustomObjectRecord> & { objectId: string; recordId: string }) =>
  apiClient.put(`/crm/custom-objects/${objectId}/records/${recordId}`, data).then((r) => r.data)

const deleteRecord = ({ objectId, recordId }: { objectId: string; recordId: string }) =>
  apiClient.delete(`/crm/custom-objects/${objectId}/records/${recordId}`).then((r) => r.data)

const addRelationship = ({
  objectId,
  recordId,
  ...data
}: Partial<CustomObjectRelationship> & { objectId: string; recordId: string }) =>
  apiClient
    .post(`/crm/custom-objects/${objectId}/records/${recordId}/relationships`, data)
    .then((r) => r.data)

const deleteRelationship = ({
  objectId,
  recordId,
  relationshipId,
}: {
  objectId: string
  recordId: string
  relationshipId: string
}) =>
  apiClient
    .delete(`/crm/custom-objects/${objectId}/records/${recordId}/relationships/${relationshipId}`)
    .then((r) => r.data)

// ─── Hooks ────────────────────────────────────────────────────────────────────

export const useCustomObjects = (params?: Record<string, unknown>) =>
  useQuery({
    queryKey: ['crm-custom-objects', params],
    queryFn: () => getCustomObjects(params),
  })

export const useCreateCustomObject = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: createCustomObject,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-objects'] }),
  })
}

export const useCustomObject = (id: string) =>
  useQuery({
    queryKey: ['crm-custom-objects', id],
    queryFn: () => getCustomObject(id),
    enabled: !!id,
  })

export const useUpdateCustomObject = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: updateCustomObject,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-objects'] }),
  })
}

export const useDeleteCustomObject = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: deleteCustomObject,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-objects'] }),
  })
}

export const useCustomObjectRecords = (objectId: string, params?: Record<string, unknown>) =>
  useQuery({
    queryKey: ['crm-custom-object-records', objectId, params],
    queryFn: () => getCustomObjectRecords(objectId, params),
    enabled: !!objectId,
  })

export const useCreateRecord = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: createRecord,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-object-records'] }),
  })
}

export const useCustomObjectRecord = (objectId: string, recordId: string) =>
  useQuery({
    queryKey: ['crm-custom-object-records', objectId, recordId],
    queryFn: () => getCustomObjectRecord(objectId, recordId),
    enabled: !!objectId && !!recordId,
  })

export const useUpdateRecord = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: updateRecord,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-object-records'] }),
  })
}

export const useDeleteRecord = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: deleteRecord,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-object-records'] }),
  })
}

export const useAddRelationship = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: addRelationship,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-object-records'] }),
  })
}

export const useDeleteRelationship = () => {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: deleteRelationship,
    onSuccess: () => qc.invalidateQueries({ queryKey: ['crm-custom-object-records'] }),
  })
}
